[For 10.4] Add share indicator in webUI

phil-davis opened the pull request:
## Description
This is a rebased and squashed version of #36254 (which has many commits) and then #36534 

Filelist shows share indicators on files and folders, that reside inside a shared folder / shared folders.
The sidebar share-tab reveals a detailed view of the share-tree including share-recipients and the respective folders the life in.

![2019-11-12-21-23-13](https://user-images.githubusercontent.com/12717530/68707852-d54a4580-0592-11ea-9585-819370243361.gif)

## Related Issue
https://github.com/owncloud/enterprise/issues/3453 (Discussion)
https://github.com/owncloud/enterprise/issues/3556 (Vis. Concept)

## How Has This Been Tested?
Manual testing in
- Google Chome
- Mozilla Firefox
- MSIE 11

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Database schema changes (next release will require increase of minor version instead of patch)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests only (no source changes)

## Checklist:
- [ ] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 
- [x] Changelog item, see [TEMPLATE](../changelog/README.md)

lordelix committed:
Add dir/path sharing info and indicator

phil-davis commented:
@individual-it I am happy with the test code (see 2nd commit). Please also give approval if you agree.

phil-davis commented:
@micbar @PVince81 please allocate someone to review the final run-time code (see the 1st commit) or review it yourself.

(the PR has had some code changes since the original code in the old PRs)

individual-it commented:
tests are fine, we can merge from my point of view

micbar commented:
@jnweiger @HanaGemela @michaelstingl @TheOneRing 
This is Frontend Only.
It doesn't affect the clients.

So ready to merge, Server QA is fine.

phil-davis commented:
Waiting on code review and approval of the run-time code.
@micbar @PVince81 who can do that and give a green tick here?

jvillafanez commented on apps/files/js/filelist.js:
```diff
@@ -1445,17 +1458,28 @@
-				if (!success) {
-					self.changeDirectory(currentDir, true);
+
+			return new Promise(function(resolve, reject) {
```
we should include in the docs that the method returns a Promise

jvillafanez commented on apps/files/js/filelist.js:
```diff
@@ -1445,17 +1458,28 @@
+					resolve()
+				});
+
+				self._updateShareTree().then(function() {
```
Is this completely independent of the "reload" promise? If the reload fails, should we update the share tree?
In addition, do we have the guarantee that the "_updateShareTree" will be executed after the reload? What if not?

jvillafanez commented on apps/files/js/filelist.js:
```diff
@@ -1445,17 +1458,28 @@
+			return new Promise(function(resolve, reject) {
+
+				if (!force && currentDir === targetDir) {
+					resolve()
```
there are quite a bunch of sentences without the ending ";". We should include it in order to keep consistency with the old code.

jvillafanez commented on apps/files/js/filelist.js:
```diff
@@ -1723,6 +1747,207 @@
+							resolve()
+						})
+					}
+					else {
```
Move this one line up for styling:
```
if (.....) {
  ......
} else {
  ......
}
```

jvillafanez commented on apps/files/js/filelist.js:
```diff
@@ -1723,6 +1747,207 @@
+		getPathShareInfo: function(path) {
+			if (typeof path !== 'string') {
+				console.error('getDirShareInfo(). param must be typeof string!')
+				return false
```
we should adjust the docs because it says the method returns a Promise, not false.

felixheidecke replied:
See below.

jvillafanez replied:
My point here is that the doc should say the method could return false if it's a valid value, otherwise I could think that returning false is a bug and it should return a rejected promise (which honestly, I think it could be an alternative).

I'm not saying that returning false is wrong, but it needs to be documented to clarify the expected behaviour of the method.

jvillafanez commented on apps/files/js/filelist.js:
```diff
@@ -1723,6 +1747,207 @@
+				crumbs.push(this.getDirShareInfo(pathToHere))
+			}
+
+			return Promise.all(crumbs)
```
any reason to return a promise and not just an array? I'd prefer not to use Promises if there is no async code inside.

felixheidecke replied:
`crumbs` is an array of promises that needs to resolve before the method can return its contents.

jvillafanez replied:
ok, got it. I overlooked that the `getDirShareInfo` returns Promises :+1:

jvillafanez commented on apps/files/js/filelist.js:
```diff
@@ -1723,6 +1747,207 @@
+			// Add share-tree icon to files and folders
+			// each per <tr> in the table
+
+			if (_.keys(this._shareTreeCache).length > 0)
```
always use blocks even for conditions with one sentence
```
if (condition) {
  method1();
} else {
  method2();
}
```

jvillafanez commented on core/js/apps.js:
```diff
@@ -44,6 +44,17 @@
+
+	exports.Apps.AppSidebarVisible = function($el) {
+		var $appSidebar = $el || $('#app-sidebar');
+		return !$appSidebar.hasClass('disappear');
```
any better option? just checking for a class doesn't seem too reliable to me.

felixheidecke replied:
I copied the code from Line 41, which also relies on the CSS ID. I guess it worked for years, so I didn't worry here.

phil-davis commented:
@micbar (cc @jvillafanez ) - who is the person who wrote this code in the first place, and/or who will be addressing the comments?
I "own" the PR only because I rebased/squashed.

felixheidecke commented:
@phil-davis @jvillafanez I addressed the issues. I took the liberty to directly push the small fixes to this branch.

jvillafanez commented on apps/files/js/filelist.js:
```diff
@@ -1723,6 +1747,209 @@
+							};
+							resolve();
+						});
+							}
```
Indentation here

jvillafanez commented on apps/files/js/filelist.js:
```diff
@@ -1723,6 +1747,209 @@
+			if (_.keys(this._shareTreeCache).length > 0) {
+				this.$fileList.find('tr td.filename .thumbnail:not(.sharetree-item)').addClass('sharetree-item');
+			}
+			else {
```
Move the "else" up to maintain the same code style:
```
if (....) {
  ....
} else {
  ....
}
```

jvillafanez commented on apps/files/js/filelist.js:
```diff
@@ -1723,6 +1747,209 @@
+
+					var shareWith = share.share_with_displayname;
+
+					if (share.share_type === OC.Share.SHARE_TYPE_GROUP)
```
use code blocks here too to keep the same code style:
```
if (.....) {
  method1();
} else {
  method2();
}
```

jvillafanez commented:
Minor remark on https://github.com/owncloud/core/pull/36572#discussion_r358226518 in case isn't visible (it's marked as outdated, but it's still present).

phil-davis commented:
JS test fails https://drone.owncloud.com/owncloud/core/22017/7/6

phil-davis commented:
@felixheidecke ping - are you available today?

phil-davis commented:
@felixheidecke ping again.
This is so close to being merged and we want to be able to make 10.4RC1

felixheidecke commented:
Hi guys. Sorry, I'm back. Will take a look right away.

felixheidecke commented:
Looks like some changes got lost in merging. Changed it back and works now.

phil-davis commented:
@felixheidecke JS tests failing https://drone.owncloud.com/owncloud/core/22057/7/6

felixheidecke commented:
@phil-davis sorry for my weired commits. I'll fix it right away.

phil-davis commented:
This LGTM now. We have had various reviews along the way of code and tests.
Can someone now review and hopefully give approval?
Then I will squash and merge.

phil-davis commented:
I added a changelog - it was missing

jvillafanez committed:
Minor code style fixes

jvillafanez commented:
I'll approve once the tests pass.

phil-davis commented:
> I'll approve once the tests pass.

When the JS tests pass then you may as welll approve. Then I can merge as soon as the remaining CI finishes.
